### PR TITLE
Add autofill of chat history, fix turnstile chat history

### PR DESCRIPTION
**Pull request by dqbd**
https://github.com/langchain-ai/langserve/assets/1443449/8cdae247-8197-4877-aa4c-0a4bcdf8d7d2

- [x] Handle function calls and additional kwargs in List[*Message]
- [x] Handle Tuple[message,message] alongside Tuple[role, message]

### CODE REVIEW

File: examples/widgets/server.py [hunk 1]
```diff
@@ -39,6 +40,14 @@ class ChatHistory(BaseModel):
     question: str
 
 
+class ChatHistoryMessage(BaseModel):
+    chat_history: List[BaseMessage] = Field(
+        ...,
+        extra={"widget": {"type": "chat", "input": "question", "output": ""}},
```
Comment (eyurtsev) [inline examples/widgets/server.py:46]: Can we do this instead?

```suggestion
        extra={"widget": {"type": "chat", "input": "question"}},
```

File: examples/widgets/server.py [hunk 2]
```diff
@@ -52,6 +61,11 @@ def chat_with_bot(input: Dict[str, Any]) -> Dict[str, Any]:
     }
 
 
+def chat_message_bot(input: Dict[str, Any]) -> BaseMessage:
```
Comment (eyurtsev) [inline examples/widgets/server.py:64]: Are we able to make this also work with List[BaseMessage] to support a use case when we're working with an agent that can also invoke tools?

**Comment by eyurtsev** [comment]
Is it easy to change the 2 column layout into something that looks more like a regular chat interface? i.e., 1 column of messages

I don't think we want to visualize hard-code the notion of request-response since the response could in the future consist of multiple messages, and there could be multiple entities (not just user and AI)

**Review by eyurtsev — APPROVED** [review]

**Review by eyurtsev — APPROVED** [review]

File: langserve/playground/src/App.tsx [hunk 3]
```diff
@@ -237,196 +265,207 @@ function App() {
...
+    <AppCallbackContext.Provider value={callbacks}>
+      <div className="flex items-center flex-col text-ls-black bg-gradient-to-b from-[#F9FAFB] to-[#EFF8FF] min-h-[100dvh] dark:from-[#0C111C] dark:to-[#0C111C]">
+        <div className="flex flex-col flex-grow gap-4 px-4 pt-6 max-w-[800px] w-full">
+          <h1 className="text-2xl text-left">
+            <strong>🦜 LangServe</strong> Playground
+          </h1>
+
+          {Object.keys(schemas.config).length > 0 && (
+            <div className="flex flex-col gap-3 [&:has(.content>.vertical-layout:first-child:last-child:empty)]:hidden">
+              {!isIframe && (
```
Comment (jacoblee93) [inline langserve/playground/src/App.tsx:285]: A little behind but when is this iframed?

File: langserve/playground/src/components/ChatMessageTuplesControlRenderer.tsx [hunk 4]
```diff
@@ -54,57 +66,59 @@ export const ChatMessageTuplesControlRenderer = withJsonFormsControlProps(
...
-            }}
+            onClick={() => props.handleChange(props.path, [...data, ["", ""]])}
           >
             <PlusIcon className="w-5 h-5" />
           </button>
         </div>
 
         <div className="flex flex-col gap-3 mt-1 empty:hidden">
-          {data.map(([type, content], index) => {
+          {data.map(([human, ai], index) => {
```
Comment (jacoblee93) [inline langserve/playground/src/components/ChatMessageTuplesControlRenderer.tsx:76]: I think Eugene brought this up but could be other chat representations of tuples?

**Comment by jacoblee93** [comment]
Seems good to me!

File: langserve/playground/src/components/ChatMessageTuplesControlRenderer.tsx (hunk 4 above)
Comment (dqbd) [reply langserve/playground/src/components/ChatMessageTuplesControlRenderer.tsx:76]: When discussed with Eugene, all of the existing templates using Tuple[str, str] for turnstile conversation at the moment. if we need to work with flat messages instead, I think we should point users to use List[BaseMessage] instead.

**Pull request merged**